Elastic Labs - Polkadot Agent Kit - Milestone 1

chungquantin opened the pull request:
# Milestone Delivery Checklist

- [x] The [milestone-delivery-template.md](https://github.com/w3f/Grant-Milestone-Delivery/blob/master/deliveries/milestone-delivery-template.md) has been copied and updated.
- [x] This pull request is being made by the same account as the accepted application.
- [x] I have disclosed any and all sources of reused code in the submitted repositories and have done my due diligence to meet its license requirements.
- [x] In case of acceptance, invoices must be submitted and payments will be transferred to the Polkadot AssetHub and/or fiat account provided in the application.
- [x] The delivery is according to the [Guidelines for Milestone Deliverables](https://grants.web3.foundation/docs/Support%20Docs/milestone-deliverables-guidelines).

Link to the application pull request: https://github.com/w3f/Grants-Program/pull/2549 < please fill this in with the PR number of your application.

chungquantin commented:
Dear reviewers / curators. 

In the official pull request of the application https://github.com/w3f/Grants-Program/pull/2549, I requested to receive the payment onchain, however, I would love to change to receive fiat for the first proposal (10k$) and receive the rest payment of the proposal in DOT and USDC as we had some changes in our company's operational plan. May I know the proper process to make the changes?

Lederstrumpf reviewed (changes requested):
Hello @chungquantin, apologies for the wait.

Thanks for delivering the first milestone - I've started review of your submission.

Your unit/e2e test suite passes on my side, and the integration tests pass on an M2's GPU and an RTX5090. I have not studied the test suite in detail yet though, so may have more feedback here.

In my review, I've focused on [your instructions for client side](https://cocdap.github.io/agent-docs/quickstart.html), see my instantiation here: https://github.com/Lederstrumpf/polkadot-agent-kit-testing
This does unfortunately not work yet:
```
/> pnpm install dist/src/quickstart.js
/> tsc
/> node dist/src/quickstart.js
node:internal/modules/cjs/loader:664
      throw e;
      ^

Error [ERR_PACKAGE_PATH_NOT_EXPORTED]: Package subpath './src/api/client' is not defined by "exports" in /home/lederstrumpf/polkadot/w3f-grants/evaluation-repos/polkadot-agent-kit-2549/polkadot-agent-kit-testing/node_modules/.pnpm/@polkadot-agent-kit+llm@1.0.0_@browserbasehq+sdk@2.6.0_@browserbasehq+stagehand@1.14.0__89468b8fe91534884d194ff6c58c1ff8/node_modules/@polkadot-agent-kit/core/package.json
    at exportsNotFound (node:internal/modules/esm/resolve:314:10)
    at packageExportsResolve (node:internal/modules/esm/resolve:661:9)
    at resolveExports (node:internal/modules/cjs/loader:657:36)
    at Function._findPath (node:internal/modules/cjs/loader:749:31)
    at Function._resolveFilename (node:internal/modules/cjs/loader:1387:27)
    at defaultResolveImpl (node:internal/modules/cjs/loader:1057:19)
    at resolveForCJSWithHooks (node:internal/modules/cjs/loader:1062:22)
    at Function._load (node:internal/modules/cjs/loader:1211:37)
    at TracingChannel.traceSync (node:diagnostics_channel:322:14)
    at wrapModuleLoad (node:internal/modules/cjs/loader:235:24) {
  code: 'ERR_PACKAGE_PATH_NOT_EXPORTED'
}

Node.js v22.17.0
```
Adding your other `@polkadot-agent-kit` packages published via https://www.npmjs.com/~cqtin0903 did not address this issue.
If you have any changes to the sdk that are not published yet and fix this, please publish them.

Notwithstanding, please find below my preliminary concerns, in my subjective descending order of priority, tagged by applicable sub-milestone:

#### <ins>M1.2</ins>: `transferNativeTool` only supports polkadot & westend relay chains
The published version:
``` typescript
type DescriptorsRelayType = {
    polkadot: typeof polkadot;
    west: typeof west;
};
```
Very easy fixes: Please add kusama and paseo too. I see you have added paseo relay to [your repo's implementation](https://github.com/openguild-labs/polkadot-agent-kit/blob/d80885c1036731424f2a18807223174ccbea9544/packages/common/src/chains/chains.ts#L21-L25), but not published this, yet kusama is certainly still missing there too, as are both kusama & paseo asset hubs.

I also note that, of the parachains, only bifrost (DOT), hydra, and paseo's people chain [are supported](https://github.com/elasticlabs-org/polkadot-agent-kit/blob/d80885c1036731424f2a18807223174ccbea9544/packages/common/src/chains/chains.ts#L49-L53). Is there a material technical reason you've omitted the others? To my understanding, your grant proposal's milestones are for a general XCM agent solution (in particular here M1.1b&c).

#### <ins>Unspecific</ins>: Agent can only be instantiated by providing private key
This isn't specific to a particular M1 sub-milestone, but the grant application specified that the sdk would support [agent creation using mnemonics](https://github.com/w3f/Grants-Program/blob/master/applications/polkadot_agent_kit.md#mnemonic-phrase-support). However, it seems that the interface currently only takes raw private keys as the first (non-optional) argument to the PolkadotAgentKit constructor. Please add support for instantiation using mnemonics.

#### <ins>M1.1b&c</ins>: incorrect/outdate `transferNativeTool` usage in quickstart & missing `xcmTransferNativeTool` call.
The `transferNativeTool` call requires a `KnowChainId` type arg, which doesn't reflect in the quickstart instructions. Likewise, the published version only has `transferNativeTool` and no `xcmTransferNativeTool`. Please address.

#### <ins>M1.0b</ins>: unclear/fragmented quickstart instructions
Please spell out the configuration steps for testers & users of your sdk fully.
For instance, a short stub on setting up ollama would be helpful so users unfamiliar with this process don't spend time troubleshooting unnecessarily.

It would be nice if you could create a template repo for instantiating polkadot agents that future users of your sdk could use as a starting point. Apologies if this already exists, and I missed it in your

#### <ins>M1.0c</ins>: CI doesn't cover e2e nor integration tests
Only the unit tests are within your CI suite, [at least on GH](https://github.com/elasticlabs-org/polkadot-agent-kit/blob/main/.github/workflows/ci.yaml#L47-L49).

#### <ins>M1.0c</ins>: integration tests fail when running on CPU
Please a timeout bypass, or set a reasonable timeout for slow ollama instances, in particular when running the model outside a GPU.
This isn't just a nit: I don't imagine that unless you get GPU instances for the `pnpm test:integration`, you'll be able to add these tests (successfully) in GH CI without increasing the timeout bound.

---

I'll review further once you have addressed the above issues.

For future deliveries:
Please test your setup & testing instructions fully before sending them in for evaluation.

Look forward to receiving your updates - I think this could be an exciting tool for users :)

chungquantin commented:
HI @Lederstrumpf, below is our report for the issues you mentioned above:

## M1.0a – Unable to run `@polkadot-agent-kit/sdk`
- **Issue**: The SDK could not be run because an outdated version was being used.  
- **Fix**: Updated to the latest release (`v1.1.3`).  

Find the v1.1.3 here: https://www.npmjs.com/package/@polkadot-agent-kit/sdk/v/1.1.3

---

## M1.1a – `transferNativeTool` limited to Polkadot & Westend
- **Issue**: `transferNativeTool` only supported Polkadot and Westend relay chains.  
- **Fix**: Extended support to additional networks (**Kusama** and **Paseo**).  
- **Commit**: [1e1f35b](https://github.com/elasticlabs-org/polkadot-agent-kit/commit/1e1f35bbdb672674baf2743472311d66a84e026f)

---

## M1.1b & M1.1c – Outdated usage examples & missing `xcmTransferNativeTool`
- **Issue**: Documentation contained outdated `transferNativeTool` usage and lacked an example for `xcmTransferNativeTool`.  
- **Fix**: Provided updated examples and a testing repo to demonstrate step-by-step setup.  
- **Reference**: [Testing Repository](https://github.com/CocDap/polkadot-agent-kit-testing)  
- **Note**: A template app will be released soon.

---

## M1.2 – Agent instantiation requires private key only
- **Issue**: Agents could only be instantiated using a private key.  
- **Fix**: Added support for **mnemonic-based** instantiation.  
- **Commit**: [3e1559d](https://github.com/elasticlabs-org/polkadot-agent-kit/commit/3e1559d477525dc5b99af2fe6ed5055e3bacaab3)

---

## M1.0b – Quickstart instructions unclear/fragmented
- **Issue**: Quickstart documentation was fragmented and confusing.  
- **Fix**: Documentation is being gradually improved to provide clear guidance for developers.  

---

## M1.0c – CI coverage & test stability
- **Issue A**: CI did not include E2E or integration tests.  
  - **Fix**: Added **E2E test coverage** to CI.  
  - **Commit**: [159e343](https://github.com/elasticlabs-org/polkadot-agent-kit/commit/159e343d5f5ac2312acb2236e3e42cab9d8b9ad2)  

- **Issue B**: Integration tests failed on CPU (timeouts with smoldot light client).  
  - **Fix**: Increased timeout handling to improve stability.  
  - **Note**: Some integration tests target **mainnet**, which cannot always run in CI frequently.  


Let me know if there are any issues!

chungquantin commented:
HI @Lederstrumpf we have submitted our response for the evaluation report

### 0b. Instructions for client integration do not work yet, perhaps some will once latest sdk changes are published.

Done: This is template testing for polkadot agent kit - Client Integration  https://github.com/CocDap/polkadot-agent-kit-testing
Future:  we will update the latest docs in this website https://cocdap.github.io/agent-docs/ 

### 0c. E2E & integration tests not part of CI pipeline yet

We have added the E2E on CI pipeline, because we have perfomed some integration tests on mainnet , then we dont add it into CI 

1b. Firstly: Integration tests don't test constrain test outcome by any criteria; see 2. Secondly: No parachain->parachain tests in E2E or integration tests. Thirdly want to test client integration before approving; see 0b.

PR for adding parachain and parachain testing: https://github.com/elasticlabs-org/polkadot-agent-kit/pull/85

### 1c. Cannot assess this with certainty while E2E tests only assert that something™ happened - only checks for "success" & return of mock results

PR for clear assertion: make sure the agent call correct tool call : https://github.com/elasticlabs-org/polkadot-agent-kit/pull/85

### 2. Firstly: Doesn't seem the integration test suite actually checks for successful results - only that something™ happened - ein nischt-nischt. Secondly: Missing both paseo & kusama relay & asset hub chains, as well as most parachains. Thirdly: Only three xcm test queries in integration test suite; these should at least be supplemented by equivalent tests for other parachain/relaychain pairings.

We have replied you on this comment https://github.com/w3f/Grant-Milestone-Delivery/pull/1270#issuecomment-3209638315

chungquantin commented:
Hi reviewers, cc @Lederstrumpf can you please let us know the next step? The submission has been in the review stage for more than a month and a half.

Lederstrumpf commented:
Hi @chungquantin,

Thank you for the updates to the delivery.
I appreciate in particular the improvement to the integration tests, i.e. that you now check that your tool is invoked with the correct parameters (matching the query).

### 0c. Unit tests + E2E tests guide + CI pipeline
> from @chungquatin: Issue B: Integration tests failed on CPU (timeouts with smoldot light client).
> Fix: Increased timeout handling to improve stability.

Thanks, note that even with the timeout increase, it still times out on an M2 Max CPU, so imagine it will likely do the same once in your CI, lest you get some GPU instances for the CI.

> from @chungquatin: Note: Some integration tests target mainnet, which cannot always run in CI frequently.

Noted, and understand the mainnet concern. In this case, could you please separate out the mainnet integration tests and only run the testnet integration tests within CI by default?
I'd add a flag then to toggle also running the mainnet tests, and then suggest doing this at least before new releases.

### 2. Thorough test coverage for XCM functionalities

via https://github.com/elasticlabs-org/polkadot-agent-kit/blob/b4bd2b24371c885ec0c42146039b21afcb71f03b/packages/sdk/tests/integration-tests/sdk.itest.ts#L98-L112:
``` typescript
    expect(result.output).toBeDefined();
    expect(result.intermediateSteps).toBeDefined();
    expect(result.intermediateSteps.length).toBeGreaterThan(0);
    
    const xcmTransferCall = result.intermediateSteps.find((step: any) => 
      step.action?.tool === 'xcm_transfer_native_asset'
    );
    
    expect(xcmTransferCall).toBeDefined();
    expect(xcmTransferCall.action.toolInput).toMatchObject({
      amount: '0.001',
      to: RECIPIENT,
      sourceChain: 'AssetHubWestend',
      destChain: 'Westend'
    });
```
While checking that your tool receives the correct inputs is clearly a marked improvement over the prior test structure, I still deem it crucial though that you test for the *on-chain* result - this is a financial tool moving real value, after all.
From my prior comment:
> Point 3. might seem pedantic, but since the ollama agent could also be initiating transacts other than intended by the user query transfer 0.001 WND to ${RECIPIENT} from Westend Asset Hub to Westend, I deem it crucial to test that no such unintended side-effects occurred.

To spell the above out: please check the onchain outcome of the tests. In the above example, this would be checking that
1. the sender's balance on Westend AH has decreased by 0.001 WND (+ transfer fees - feel free to just add a reasonable upper bound for these when checking the balance value)
2. the recipient's balance on Westend has increased by 0.001 WND
3. no other side-effects occurred

---

Note also that 3 of 7 integration tests fail, see attached logs from both nixos & ubuntu testing (with an RTX5090); logs below:
- [integration-RTX5090-nixos.log](https://github.com/user-attachments/files/22078320/integration-RTX5090-nixos.log)
- [integration-RTX5090-remote-ubuntu.log](https://github.com/user-attachments/files/22078319/integration-RTX5090-remote-ubuntu.log)

For example:
```bash
@polkadot-agent-kit/sdk:test:integration: AssertionError: expected undefined to be defined
@polkadot-agent-kit/sdk:test:integration:  ❯ tests/integration-tests/sdk.itest.ts:161:22
@polkadot-agent-kit/sdk:test:integration:     159|     expect(checkBalanceCalls.length).toBeGreaterThanOrEqual(1);
@polkadot-agent-kit/sdk:test:integration:     160|
@polkadot-agent-kit/sdk:test:integration:     161|     expect(initCall).toBeDefined();
@polkadot-agent-kit/sdk:test:integration:        |                      ^
@polkadot-agent-kit/sdk:test:integration:     162|     expect(initCall.action.toolInput).toEqual({
@polkadot-agent-kit/sdk:test:integration:     163|       chainId: 'polkadot'
```

> For future deliveries:
> Please test your setup & testing instructions fully before sending them in for evaluation.

Reiterating this: please test your instructions on a clean system (and run all tests) before sending in for evaluation.
In addition to the failing test above, the delivery lacked a package import, see https://github.com/elasticlabs-org/polkadot-agent-kit/pull/92

chungquantin commented:
Hi @Lederstrumpf, can you test it again?

CocDap commented:
### **1. Problem: Root Cause of Test Case Failure**

I've identified the root cause for the failure of the `should call xcm_transfer_native_asset tool for Westend to Asset Hub transfer` test case.

The issue stems from the `qwen3:latest` model's inability to correctly map internal chain IDs to their corresponding chain names. Despite being provided with clear conversion rules in the prompt https://github.com/elasticlabs-org/polkadot-agent-kit/blob/fix/integration/packages/llm/src/prompt/index.ts#L113) , the model confuses the chain names, leading to an incorrect tool call and subsequent failure.

### **2. Solution: Specialized Prompts**

To resolve this, I have refactored the prompts. Instead of a single, large prompt for all tools, I have created smaller, specialized prompts for each distinct task (e.g., `assets`, `xcm`). This reduces ambiguity and leads to more accurate tool calls, especially for less advanced models.

### **3. Evidence: Model Performance Comparison**

Link Testing: https://github.com/CocDap/polkadot-agent-kit-testing

To demonstrate the reasoning limitations of `qwen3:latest` compared to more advanced models, I created a test script. The results clearly show the difference:

- **With Ollama `qwen3:latest`:**
    - **Query:**
    
    ```jsx
    runAgent("transfer 0.1 WND to 5Ccmxb84eREZmtSkrLJSYp6QxJwNvmNbrfBm4p5B5VnKrB8z from Westend to West Asset Hub");
    ```
    
    - **Result:** **Failure**
    
    ```jsx
    - Tool Result (xcm_transfer_native_asset): {"content":"{\"success\":false,\"error\":{\"code\":1005,\"message\":\"Unknown error occurred\",\"name\":\"GenericError\"},\"tool\":\"xcm_transfer_native_asset\",\"timestamp\":\"2025-09-26T08:38:40.061Z\"}","tool_call_id":"xcm_transfer_native_asset_error_1758875920061"}
    ```
    
- **With `gemini-2.0-flash`:**
    - **Query:**
    
    ```jsx
    runAgent("transfer 0.1 WND to 5Ccmxb84eREZmtSkrLJSYp6QxJwNvmNbrfBm4p5B5VnKrB8z from Westend to West Asset Hub", "gemini");
    ```
    
    - **Result:** **Success**
    
    ```jsx
    - Tool Result (xcm_transfer_native_asset): {"content":"{\"success\":true,\"data\":\"Tx Hash Successful: 0xef3e18771dcd73baea22fc1fd0e8588e4a90f9fec90e67d33c04865145622eee\",\"tool\":\"xcm_transfer_native_asset\",\"timestamp\":\"2025-09-26T08:31:11.293Z\"}","tool_call_id":"xcm_transfer_native_asset_1758875471293"}
    ```
    

This confirms that while the user's query is correct, a less capable model can produce an incorrect outcome. The solution for these models is to use more specific, less complex prompts.

### **4. Additional Improvements**

- **Enhanced Test Cases:** I really appreciate your doubt, then I have updated the test suite with more case of  user intents for the same action (e.g., "Westend to Westend Asset Hub", "Westend's relay chain to Westend's Asset Hub") to ensure the agent correctly handles variations in phrasing. (https://github.com/elasticlabs-org/polkadot-agent-kit/blob/fix/integration/packages/sdk/tests/integration-tests/sdk.itest.ts#L218))

Link PR: https://github.com/elasticlabs-org/polkadot-agent-kit/pull/109


- **Hardware Limitations:** Regarding testing on other LLM hosts (Nvidia/AMD GPUs, CPU Ollama), we currently do not have the necessary hardware resources to perform these tests.

### **5. Noted:**

Kindly checking the integration tests on this PR (https://github.com/elasticlabs-org/polkadot-agent-kit/pull/109) 

**Command to run :**

```
pnpm run test:integration:testnet
```

### **6. Conclusion**
I've spammed the tests a lot, and all were successful. You can check the results on Subscan with this wallet: `5CwaVNAmJ6hWvu9tgxAPgxLD27SK36Vk4hgDcSZGK7z2eYiE`

<img width="1655" height="764" alt="image" src="https://github.com/user-attachments/assets/bb03968c-b2d4-44b2-9e81-bfed75d1aed6" />

CC: @chungquantin @Lederstrumpf 

PS: Thank you for your careful review and thoughtful suggestion

Lederstrumpf reviewed (changes requested):
Hi @CocDap (cc. @chungquantin),

Thank you for the updates, and apologies for my delayed re-review.
I appreciate the detailed explanation+justification you provided, the increased testing scope of phrasing variations, and that the instructions for review in your latest comment were much clearer.

The new submission unfortunately also doesn't work: wrt testnet tests, potentially for me, but for the mainnet test, certainly in general too.
Given my delay in responding, I'll give more details as to what's failing than I intended to:

1. I recognize that the testnet integration tests must have passed for you (assuming on https://github.com/elasticlabs-org/polkadot-agent-kit/pull/109 / https://github.com/elasticlabs-org/polkadot-agent-kit/commit/ca8ba5843f6c89084318f6f00d5b01d3a80aab23 based on the test count in the screenshot), but they don't for me, either on  https://github.com/elasticlabs-org/polkadot-agent-kit/commit/ca8ba5843f6c89084318f6f00d5b01d3a80aab23 or its two followup commits. Please find below logs of all the failures I encountered in said revisions at the bottom of this comment. When exactly did you run the tests again? Asset Hub migration for Paseo was ~ a month before your resubmission (https://forum.polkadot.network/t/asset-hub-migration-2025/11129/71), so I don't believe this is related.

2. More critical than the testnet integration test failure, the mainnet test doesn't work, and given the changes made in https://github.com/elasticlabs-org/polkadot-agent-kit/pull/109, I must assume you did not run this test again before requesting my re-review.

3. The correction I pushed for 2. is only the first step. Please also test the observation of that test: the observation `observation: {"content":"{\\"success\\":true,\\"data\\":\\"Cross-chain swap failed: 0.1 DOT to USDt from polkadot to polkadot_asset_hub. Error: Cannot read properties of undefined (reading 'otherAssets')\\",\\"tool\\":\\"swap_tokens\\",\\"timestamp\\":\\"2025-10-15T21:11:57.842Z\\"}","tool_call_id":"swap_tokens_1760562717842"}` is currently passing and obviously should not.

4. Please also extend the testing for mainnet operation - that [single DOT/USDt swap test](https://github.com/elasticlabs-org/polkadot-agent-kit/blob/main/packages/sdk/tests/integration-tests/sdk.mainnet.itest.ts#L27-L55) is not sufficient (very narrow scope) - I'd expect something comparable to the testnet scope - especially given your recent move to a more "handwired" solution with context-specialized prompts, as you noted in https://github.com/w3f/Grant-Milestone-Delivery/pull/1270#issuecomment-3350097596.

------

Test logs of 1.:
- [testnet-aba3953f4f45f385b4617af348e1fc2df7d88f22.log](https://github.com/user-attachments/files/22936979/testnet-aba3953f4f45f385b4617af348e1fc2df7d88f22.log) (current `main`)
 - [testnet-cbb7e92a484be65df1f2467371bf0e74fad043f3.log](https://github.com/user-attachments/files/22936983/testnet-cbb7e92a484be65df1f2467371bf0e74fad043f3.log) (current `main~1`)
 - [testnet-ca8ba5843f6c89084318f6f00d5b01d3a80aab23.log](https://github.com/user-attachments/files/22936985/testnet-ca8ba5843f6c89084318f6f00d5b01d3a80aab23.log) (current `main~2`, i.e. https://github.com/elasticlabs-org/polkadot-agent-kit/pull/109)

CocDap commented:
To provide proof for my assertion, I manually observed the balances, and they correctly match the asserted `before`  and `after` states

```typescript
    expect(balanceRecipientAfter.data.free).toBeLessThan(balanceRecipientBefore.data.free + amountParsed);

    expect(balanceAgentAfter.data.free).toBeLessThan(balanceAgentBefore.data.free - amountParsed - feeXCM.fee);
```

My manual check: 
```
Before Balance of Recipient on Paseo
{
  nonce: 2
  consumers: 0
  providers: 1
  sufficients: 0
  data: {
    free: 85,290,420,868
    reserved: 0
    frozen: 0
    flags: 170,141,183,460,469,231,731,687,303,715,884,105,728
  }
}

Before Balance of Agent on Paseo Asset Hub 
{
  nonce: 281
  consumers: 1
  providers: 2
  sufficients: 0
  data: {
    free: 49,521,302,891,008
    reserved: 432,419,832,838
    frozen: 0
    flags: 170,141,183,460,469,231,731,687,303,715,884,105,728
  }
}

Amount Transfer  : 0.1 PAS -> 1000000000

Agent Transfer 0.1 PAS to RECIPIENT from Paseo Asset Hub to Paseo 


Tx Hash : 0xbaba9ad237cfdf6ad83132111f948db159f452e341b068709126f66d843f19db


After Balance of Recipient on Paseo

{
  nonce: 2
  consumers: 0
  providers: 1
  sufficients: 0
  data: {
    free: 86,266,959,406
    reserved: 0
    frozen: 0
    flags: 170,141,183,460,469,231,731,687,303,715,884,105,728
  }
}

After Balance of Agent on Paseo Asset Hub 

{
  nonce: 282
  consumers: 1
  providers: 2
  sufficients: 0
  data: {
    free: 49,519,982,023,644
    reserved: 432,419,832,838
    frozen: 0
    flags: 170,141,183,460,469,231,731,687,303,715,884,105,728
  }
}

My Integration Tests Assertion to check after and before balance 

balance recipient after < balance recipient before + amount 
86,266,959,406 < 85,290,420,868 + 1000000000(86,290,420,868) -> correct 


balance agent after < balance agent before - amount - xcm fee 
49,519,982,023,644 < 49,521,302,891,008 - 1000000000 - 18375000 (49,520,284,516,008) -> correct 
```
I trust that the core Parity team has already thoroughly tested the before-and-after balance logic also

Lederstrumpf commented:
Hi @CocDap, 

Hopefully my at-most penultimate followup here ;)

> 2. Clarification on Mainnet Swap Tests
> I saw your notes on the mainnet swap tests failing. I am aware of this issue. However, please note that this functionality is explicitly a requirement for Milestone 2, not Milestone 1.
> I will address these swap-related issues in a separate PR as part of the work for the next milestone.

Milestone 2 does not mention mainnet tests anywhere and M1.2 promises integration tests with "Thorough test coverage for XCM functionalities" already. So I really read this as being part of milestone 1.
But I'm fine if you only address the mainnet failures in milestone 2 - the timing is not so important since it seems clear to me you'll deliver M2 as well.

> 3. Clarification on Failures in the main Branch

Acknowledged - and the nomination pools are indeed only part of milestone 2.

---

Please find my updated evaluation over here:
You'll note that it now accepts all deliveries except for M1.0b

>M1.0b – Quickstart instructions unclear/fragmented
>
>   Issue: Quickstart documentation was fragmented and confusing.
>   Fix: Documentation is being gradually improved to provide clear guidance for developers.

Unfortunately, the docs linked in your PR (https://cocdap.github.io/agent-docs/quickstart.html) have still not been updated - the latest state is still from July: https://github.com/elasticlabs-org/polkadot-agent-kit-docs/commit/a97343d1ce49f19260f0ec8bbffdaf2eb5c86383

And your interfaces have changed quite a bit in the meantime - you'll see this once you try running those instructions.
So: please update the quickstart instructions so that they allow a developer to ... well.. get started quickly.
I reckon it makes sense to reuse parts of https://github.com/CocDap/polkadot-agent-kit-testing for this.

Once you update the quickstart instructions and they all work smoothly in my review, I'm happy to approve your PR.

CocDap commented:
That's great news, thank you for the approval! We just wanted to add a special thanks for the high-quality review. Your @Lederstrumpf  feedback was clear, fair, and has definitely helped us improve.
❤️ 
https://github.com/w3f/Grant-Milestone-Delivery/pull/1293
CC: @chungquantin

CocDap commented:
I really appreciate your detailed feedback. The SDK has become significantly more user-friendly and maintainable
